### PR TITLE
feat/#2338: support debug multi goroutine

### PR DESCRIPTION
**Pull request by xbsura**
Now vim-go support  multi goroutine debug

Add two command: GoDebugGoroutines to show all goroutines running, GoDebugGoroutine + goroutineId to switch between goroutines

with this request: #2338

### CODE REVIEW

File: autoload/go/debug.vim [hunk 1]
```diff
@@ -899,6 +911,49 @@ function! s:isActive()
...
+    try
+      let l:currentGoroutineId = s:groutineID()
+    catch 
+      echom "current goroutineId not found..."
+    endtry
+
+    let l:res = s:call_jsonrpc('RPCServer.ListGoroutines')
+    let l:goroutines = l:res.result.Goroutines
+    if len(l:goroutines) == 0
+      echom "No Goroutines Running Now..."
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Can you change `echom` here to `go#util#EchoWarning`?

File: autoload/go/debug.vim [hunk 2]
```diff
@@ -899,6 +911,49 @@ function! s:isActive()
...
 endfunction
 
+" List All Goroutines Running
+function! go#debug#Goroutines(...) abort
+  try
+    let l:currentGoroutineId = 0
+    try
+      let l:currentGoroutineId = s:groutineID()
+    catch 
+      echom "current goroutineId not found..."
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Can you change `echom` here to `go#util#EchoWarning`?

File: autoload/go/debug.vim [hunk 3]
```diff
@@ -899,6 +911,49 @@ function! s:isActive()
...
+    endtry
+endfunction
+
+" Change Goroutine
+function! go#debug#Goroutine(goroutineId) abort
+  let l:goroutineId = a:goroutineId
+  try
+    let l:res = s:call_jsonrpc('RPCServer.Command', {'Name': 'switchGoroutine', 'GoroutineID': str2nr(l:goroutineId)})
+    call s:stack_cb(l:res)
+    echom "Switched Goroutine to: " . l:goroutineId
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Can you change `echom` here to `go#util#EchoInfo`?

File: autoload/go/debug.vim [hunk 4]
```diff
@@ -899,6 +911,49 @@ function! s:isActive()
...
+    let l:goroutines = l:res.result.Goroutines
+    if len(l:goroutines) == 0
+      echom "No Goroutines Running Now..."
+      return
+    endif
+
+    for l:idx in range(len(l:goroutines))
+      let l:goroutine = l:goroutines[l:idx]
+      if l:goroutine.id == l:currentGoroutineId
+        echom "* Goroutine " . l:goroutine.id . " - " . l:goroutine.userCurrentLoc.file . ":" . l:goroutine.userCurrentLoc.line .  " " l:goroutine.userCurrentLoc.function.name . " " . " (thread: " . l:goroutine.threadID . ")"
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Can you change `echom` here to `go#util#EchoInfo`?

File: autoload/go/debug.vim [hunk 5]
```diff
@@ -899,6 +911,49 @@ function! s:isActive()
...
+      else
+        echom "  Goroutine " . l:goroutine.id . " - " . l:goroutine.userCurrentLoc.file . ":" . l:goroutine.userCurrentLoc.line .  " " l:goroutine.userCurrentLoc.function.name . " " . " (thread: " . l:goroutine.threadID . ")"
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Can you change echom here to go#util#EchoInfo?

**Comment by xbsura** [comment]
> Thank you for contributing!. This is good work, and I look forward to integrating it.
> 
> I think, though, that probably makes sense to replace the new `:GoDebugGoroutines` with a window that lists all the goroutines. It should clearly indicate the active goroutine with a preceding `*`.
> 
> What do you think?

yes, current goroutine has a * now

File: autoload/go/debug.vim (hunk 1 above)
Comment (xbsura) [reply autoload/go/debug.vim]: fixed

File: autoload/go/debug.vim (hunk 2 above)
Comment (xbsura) [reply autoload/go/debug.vim]: fixed

File: autoload/go/debug.vim (hunk 3 above)
Comment (xbsura) [reply autoload/go/debug.vim]: fixed

File: autoload/go/debug.vim (hunk 4 above)
Comment (xbsura) [reply autoload/go/debug.vim]: fixed

File: autoload/go/debug.vim (hunk 5 above)
Comment (xbsura) [reply autoload/go/debug.vim]: fixed

**Comment by xbsura** [comment]
HI:

Now a new window is created when running debug, and I resize some window to make it more resonable, and type enter on goroutine can switch between goroutines just like that:

image

File: autoload/go/config.vim [hunk 6]
```diff
@@ -205,9 +205,10 @@ endfunction
 
 function! go#config#DebugWindows() abort
   return get(g:, 'go_debug_windows', {
-            \ 'stack': 'leftabove 20vnew',
-            \ 'out':   'botright 10new',
-            \ 'vars':  'leftabove 30vnew',
+            \ 'goroutines':  'split 60vnew',
+            \ 'out':   'botright 100new',
```
Comment (bhcleek) [inline autoload/go/config.vim]: did you mean to change this from 10 to 100?

File: autoload/go/debug.vim [hunk 7]
```diff
@@ -445,6 +455,13 @@ function! s:start_cb() abort
     nmap <buffer> q <Plug>(go-debug-stop)
   endif
 
+  call win_gotoid(l:winid)
+  exe "resize +12"
+  if l:goroutine_win_id != 0
+    call  win_gotoid(l:goroutine_win_id)
+    exe "resize +6"
+  endif
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Why does this resize windows?

File: autoload/go/debug.vim [hunk 8]
```diff
@@ -771,6 +788,68 @@ function! go#debug#Print(arg) abort
...
 endfunction
 
+function! s:update_goroutines() abort
+  let l:var_win = bufwinnr(bufnr('__GODEBUG_GOROUTINES__'))
+  if l:var_win == -1
+    return
+  endif
+
+  let l:cur_win = bufwinnr('')
+  exe l:var_win 'wincmd w'
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Use `win_gotoid` instead of normal mode commands.

File: autoload/go/debug.vim [hunk 9]
```diff
@@ -771,6 +788,68 @@ function! go#debug#Print(arg) abort
...
+      else
+        let l:g = "  Goroutine " . l:goroutine.id . " - " . l:goroutineType . ": " . l:loc.file . ":" . l:loc.line .  " " . l:loc.function.name . " " . " (thread: " . l:goroutine.threadID . ")"
+      endif
+      let v += [l:g]
+    endfor
+
+    call setline(1, v)
+  finally
+    setlocal nomodifiable
+    exe l:cur_win 'wincmd w'
```
Comment (bhcleek) [inline autoload/go/debug.vim]: Use `win_gotoid`

**Comment by xbsura** [comment]
> This is looking really great. Thank you for the contribution. There's a couple of changes that I'd like to see, but they're nitpicks, so I can do them later if you'd prefer to be done with this. I would appreciate it, though, if you could explain the rationale for why the windows are resized in `s:start_cb`.

be free to make any change

the default size of window make code window too small, so I make the code window bigger and make goroutine window smaller

**Comment by bhcleek** [comment]
I have a local branch that fixes the tests, cleans up some code, and make some other simple changes, too. Do you mind if I push to your branch?

**Comment by xbsura** [comment]
be free to make any change, just push your code

**Pull request merged**